### PR TITLE
Key-only deleters in Storage.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: Storage/storage.h [hunk 1]
```diff
@@ -78,11 +76,14 @@ namespace storage {
 #define CURRENT_STORAGE_STRUCT_TAG(base, alias)                                               \
   CURRENT_STRUCT(CURRENT_STORAGE_ADDER_##alias, base) {                                       \
     CURRENT_DEFAULT_CONSTRUCTOR(CURRENT_STORAGE_ADDER_##alias) {}                             \
-    CURRENT_CONSTRUCTOR( CURRENT_STORAGE_ADDER_##alias)(const base& value) : SUPER(value) {}  \
+    CURRENT_CONSTRUCTOR(CURRENT_STORAGE_ADDER_##alias)(const base& value) : SUPER(value) {}   \
   };                                                                                          \
-  CURRENT_STRUCT(CURRENT_STORAGE_DELETER_##alias, base) {                                     \
+  CURRENT_STRUCT(CURRENT_STORAGE_DELETER_##alias) {                                           \
+    CURRENT_FIELD(key, ::current::storage::sfinae::ENTRY_KEY_TYPE<base>); \
```
Comment (dkorolev) [inline Storage/storage.h]: `s/key/key_to_erase/` ?

File: Storage/storage.h [hunk 2]
```diff
@@ -78,11 +76,14 @@ namespace storage {
...
     CURRENT_DEFAULT_CONSTRUCTOR(CURRENT_STORAGE_DELETER_##alias) {}                           \
-    CURRENT_CONSTRUCTOR(CURRENT_STORAGE_DELETER_##alias)(const base& value) : SUPER(value) {} \
+    CURRENT_CONSTRUCTOR(CURRENT_STORAGE_DELETER_##alias)(const base& value) { \
```
Comment (dkorolev) [inline Storage/storage.h]: Why assignment, not initialization?

File: Storage/test.cc [hunk 3]
```diff
@@ -48,6 +48,8 @@ namespace transactional_storage_test {
 CURRENT_STRUCT(Element) {
   CURRENT_FIELD(x, int32_t);
   CURRENT_CONSTRUCTOR(Element)(int32_t x = 0) : x(x) {}
+  // Dummy key function to support key-only deleter.
+  int32_t key() const { return x; }
```
Comment (dkorolev) [inline Storage/test.cc]: Shouldn't `Element` already have it own key field/method defined to be used in the storage?

On the general note, why not have some `CURRENT_PRIMARY_KEY(...);` (and `CURRENT_PRIMARY_KEY_METHOD()` macro, much like we have `CURRENT_TIMESTAMP()`?

File: Storage/storage.h (hunk 1 above)
Comment (mzhurovich) [inline Storage/storage.h]: For two reasons:
1) It's shorter
2) `map_.erase(sfinae::GetKey(object));` works out of the box :)

File: Storage/storage.h (hunk 2 above)
Comment (mzhurovich) [inline Storage/storage.h]: Just thought it's more readable.

File: Storage/test.cc (hunk 3 above)
Comment (mzhurovich) [inline Storage/test.cc]: Speaking of must-have key field - generally yes, but this structure is used in Vector, which doesn't need keys. Are we going to deprecate it?
In general, we'll need composite keys for matrices. So, let's return to macro design question when we implement `ManyToMany` etc.

File: Storage/storage.h (hunk 1 above)
Comment (dkorolev) [inline Storage/storage.h]: Sold :+1:

File: Storage/storage.h (hunk 2 above)
Comment (dkorolev) [inline Storage/storage.h]: In the future far, far away, `key` might not be default-constructible.

File: Storage/test.cc (hunk 3 above)
Comment (dkorolev) [inline Storage/test.cc]: If the issue is with `Vector` only, which is going away, I'm fine not thinking of it now.

File: Storage/storage.h (hunk 2 above)
Comment (mzhurovich) [inline Storage/storage.h]: Ok, changed to init-list.

File: Storage/test.cc (hunk 3 above)
Comment (mzhurovich) [inline Storage/test.cc]: Yep, it's only for Vector.

**Comment by dkorolev** [comment]
LGTM!

**Pull request merged**